sandbox/cgroup/tracking_test.go

```diff
@@ -0,0 +1,580 @@
+	defer restore()
+
+	// Calling StartTransientUnit is not attempted without a DBus connection.
+	restore = cgroup.MockDoCreateTransientScope(func(conn *dbus.Conn, unitName string, pid int) error {
+		panic("we are not expecting this call")
+	})
+	defer restore()
+
+	// Disable the cgroup analyzer function as we don't expect it to be used in this test.
+	restore = cgroup.MockCgroupProcessPathInTrackingCgroup(func(pid int) (string, error) {
+		panic("we are not expecting this call")
+	})
+	defer restore()
+
+	// Attempts to create a transient scope fail with a special error
+	// indicating that we cannot track application process.
+	err := cgroup.CreateTransientScopeForTracking("snap.pkg.app")
+	c.Assert(err, ErrorMatches, "cannot track application process")
+}
+
+func (s *trackingSuite) TestCreateTransientScopeForTrackingSilentlyFails(c *C) {
+	// Pretend that refresh app awareness is enabled
+	enableFeatures(c, features.RefreshAppAwareness)
+
+	// Hand out stub connections to both the system and session bus.
+	// Neither is really used here but they must appear to be available.
+	restore := dbusutil.MockConnections(dbustest.StubConnection, dbustest.StubConnection)
+	defer restore()
+
+	// Pretend we are a non-root user.
+	restore = cgroup.MockOsGetuid(12345)
+	defer restore()
+	// Pretend our PID is this value.
+	restore = cgroup.MockOsGetpid(312123)
+	defer restore()
+
+	// Rig the random UUID generator to return this value.
+	uuid := "cc98cd01-6a25-46bd-b71b-82069b71b770"
+	restore = cgroup.MockRandomUUID(uuid)
+	defer restore()
+
+	// Calling StartTransientUnit succeeds but in reality does not move our
+	// process to the new cgroup hierarchy. This can happen when systemd
+	// version is < 238 and when the calling user is in a hierarchy that is
+	// owned by another user. One example is a user logging in remotely over
+	// ssh.
+	restore = cgroup.MockDoCreateTransientScope(func(conn *dbus.Conn, unitName string, pid int) error {
+		return nil
+	})
+	defer restore()
+
+	// Rig the cgroup analyzer to pretend that we are not placed in a snap-related slice.
+	restore = cgroup.MockCgroupProcessPathInTrackingCgroup(func(pid int) (string, error) {
+		c.Assert(pid, Equals, 312123)
+		return "/system.slice/foo.service", nil
+	})
+	defer restore()
+
+	// Attempts to create a transient scope fail with a special error
+	// indicating that we cannot track application process even though
+	// the DBus call has returned no error.
+	err := cgroup.CreateTransientScopeForTracking("snap.pkg.app")
+	c.Assert(err, ErrorMatches, "cannot track application process")
+}
+
+func checkAndRespondToStartTransientUnit(c *C, msg *dbus.Message, scopeName string, pid int) *dbus.Message {
+	// XXX: Those types might live in a package somewhere
+	type Property struct {
+		Name  string
+		Value interface{}
+	}
+	type Unit struct {
+		Name  string
+		Props []Property
+	}
+	// Signature of StartTransientUnit, string, string, array of Property and array of Unit (see above).
```

Comment:
by above did you mean the other package ?

Reply:
(if so, perhaps it would be useful to point to that package / function comment or otherwise copy relevant bits of the signature we are using here)

Reply:
No, literally above this line.